Shared Regs functional with test programs

ThomasSeaver opened the pull request:
Altered VirtualMachine.cs to extend reg num, add propagation call to data manipulation instructions, and add method to update shared regs from within vm

Altered BotManager.cs to add coroutine that, after a delay, uses the method in bot.VM to update shared regs across teams

Added SharedRegisters and SharedRegisters2 as testing programs that demonstrate the bots waiting for eachother to update the regs. You can examine the programs yourself, but they basically are taking turns changing the reg and then waiting for their partner to change the reg, moving each time they set a reg for visual clarity.

I did not edit the editor's display of regs since that is probably something we should discuss beforehand and the system extends to 10 regs automatically via the numreg change. 

I exposed the propagation delay variable in the inspector, but the coroutine seems to use the first value in the variable even after changing it in the inspector? I'm really not sure what's going on there; even after changing it in the inspector, the debug log would report the original value. I'm not even sure it was using the inspector value at all, and I think I would reset to 0 unless I set the delay manually in the awake function. Regardless, the delay functions, the coroutine just doesn't seem to want to use any updated version of the variable, and I'm not sure what part of C# I'm running up against there.

ThomasSeaver committed:
Shared Regs functional with test program

Altered VirtualMachine.cs to extend reg num, add propagation call to data manipulation instructions, and add method to update shared regs from within vm

Altered BotManager.cs to add coroutine that, after a delay, uses the method in bot.VM to update shared regs across teams

Added SharedRegisters and SharedRegisters2 as testing programs that demonstrate the bots waiting for eachother to update the regs.

BrandonShockley commented on Assets/Decontainment/Singletons/BotManager.cs:
```diff
@@ -71,4 +75,14 @@ public int GetTargetHeading(Controller targeter, int targetIndex)
         return (int)Vector2.SignedAngle(targeter.transform.right, look);
     }
+
+    public IEnumerator PropagationCoroutine (int registerNumber, int registerValue, int botTeamID) {
```
```suggestion
    public IEnumerator PropagationCoroutine(int registerNumber, int registerValue, int botTeamID)
    {
```
Stay consistent with the rest of the codebase

BrandonShockley replied:
I highly-highly recommend making the coroutine method private and making a `public void PropagateRegister` method that sets up the coroutine for you. That way you don't have to type as much code to do the propagation like you did in the VirtualMachine class.

BrandonShockley commented on Assets/Decontainment/Assembly/VirtualMachine.cs:
```diff
@@ -164,6 +171,10 @@ public void Tick()
         OnTick?.Invoke();
     }
 
+    public void updateSharedReg(int registerNumber, int registerValue) {
```
```suggestion
    public void UpdateSharedReg(int registerNumber, int registerValue)
    {
```
Keep the code style consistent

BrandonShockley replied:
Also this method isn't used anywhere

BrandonShockley commented on Assets/Decontainment/Singletons/BotManager.cs:
```diff
@@ -71,4 +75,14 @@ public int GetTargetHeading(Controller targeter, int targetIndex)
     }
+
+    public IEnumerator PropagationCoroutine (int registerNumber, int registerValue, int botTeamID) {
+        //Debug.Log("Propagating with delay: " + propagationDelay);
```
Don't leave commented out code in the PR

BrandonShockley commented on Assets/Decontainment/Singletons/BotManager.cs:
```diff
@@ -71,4 +75,14 @@ public int GetTargetHeading(Controller targeter, int targetIndex)
+
+    public IEnumerator PropagationCoroutine (int registerNumber, int registerValue, int botTeamID) {
+        //Debug.Log("Propagating with delay: " + propagationDelay);
+        yield return new WaitForSecondsRealtime(propagationDelay);
```
Change this to use WaitForSeconds

BrandonShockley commented on Assets/Decontainment/Singletons/BotManager.cs:
```diff
@@ -15,6 +16,8 @@ public enum DistanceType
         FARTHEST,
     }
 
+    public float propagationDelay;
```
Make this a private variable with the `[SerializeField]` attribute

BrandonShockley commented on Assets/Decontainment/Assembly/VirtualMachine.cs:
```diff
@@ -6,7 +6,7 @@
 public class VirtualMachine
 {
-    public const int NUM_REGS = 5;
+    public const int NUM_REGS = 10;
```
Do what we discussed in the chat

ThomasSeaver committed:
Updating with requested changes and editor support

ThomasSeaver committed:
Merge master onto updated feature

ThomasSeaver committed:
Merge master completely

ThomasSeaver committed:
Modifying color slightly

ThomasSeaver commented:
Updated as requested in regards to style, and added new register group to the editor with new color as to visually differentiate. Still functional as before.

BrandonShockley commented on Assets/Decontainment/Singletons/BotManager.cs:
```diff
@@ -26,6 +30,7 @@ public enum DistanceType
 
     void Awake()
     {
+        propagationDelay = 3.0f;
```
Remove this

ThomasSeaver committed:
Singleton change and propagation delay removal
